### ISSUE TITLE
REALM_JNI: flock() failed: Bad file descriptor in io_realm_internal_OsSharedRealm.cpp

### ISSUE THREAD

**Issue by mariusboepple** [realm/realm-java#5908]
We have been struggling with more or less random errors for some time now. Lately I've experienced this one during developement:

```
04-18 12:51:32.263 E/REALM_JNI: jni: ThrowingException 5, flock() failed: Bad file descriptor in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_OsSharedRealm.cpp line 126, .
    Exception has been thrown: Unrecoverable error. flock() failed: Bad file descriptor in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_OsSharedRealm.cpp line 126
04-18 12:51:32.270 E/AndroidRuntime: FATAL EXCEPTION: SyncAdapterThread-3
    Process: com.example.app, PID: 12352
    io.realm.exceptions.RealmError: Unrecoverable error. flock() failed: Bad file descriptor in /Users/cm/Realm/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_OsSharedRealm.cpp line 126
        at io.realm.internal.OsSharedRealm.nativeBeginTransaction(Native Method)
        at io.realm.internal.OsSharedRealm.beginTransaction(OsSharedRealm.java:269)
        at io.realm.BaseRealm.beginTransaction(BaseRealm.java:401)
        at io.realm.Realm.beginTransaction(Realm.java:146)
        at io.realm.Realm.executeTransaction(Realm.java:1403)
        at com.example.app.sync.SyncAdapter.syncCars(SyncAdapter.kt:1266)
        at com.example.app.sync.SyncAdapter.syncComplete(SyncAdapter.kt:140)
        at com.example.app.sync.SyncAdapter.onPerformSync(SyncAdapter.kt:101)
        at android.content.AbstractThreadedSyncAdapter$SyncThread.run(AbstractThreadedSyncAdapter.java:321)
```

> Note: The sync adapter framework runs onPerformSync() on a background thread, so you don't have to set up your own background processing.

The code of the `SyncAdapter` wasn't changed for quite a while. At the specific line, there is just an {{realm.executeTransaction { ... }}}.

Can you guys read anything from the stack trace? What's causing this crash?

#### Version of Realm and tooling
Realm version(s): 5.0.0

Realm sync feature enabled: no

Android Studio version: 3.1.1

Which Android version and device: Pixel 2 (Android 8.1.0)

**Comment by cmelchior** [comment]
Interesting. The Bad file descriptor can happen for multiple reasons, but I don't think it can happen due to anything you do. So perhaps it is a race condition somewhere on our end.

A couple of questions:
- Did you recently upgrade Realm or is the error just starting to happen
- Are you deleting the Realm file manually anywhere?
- How are you getting the Realm instance in the SyncAdapter?

**Comment by MariusBoepple** [comment]
> Did you recently upgrade Realm or is the error just starting to happen

I've updated from `4.3.3` to `5.0.0` in my development branch. We have no reports of this particular crash and I've only had it one time during development.

> Are you deleting the Realm file manually anywhere?

We run the following code when the user changed. But I think I did not change my account during development:
    
    // Check if the account changed.
    if (oldName != null && oldName != account.name) {
            Log.w("Clearing data because account has changed from [$oldName] to [${account.name}]")
            Realm.deleteRealm(realmConfig)
            SharedPreferencesUtils.clear(context)
    }

Besides that, we're not deleting anything.

> How are you getting the Realm instance in the SyncAdapter?

    internal val realmConfig: RealmConfiguration by lazy {
        val builder = RealmConfiguration.Builder()
            .name(REALM_NAME_V4)
            .modules(MyRealmModule())
            .schemaVersion(REALM_SCHEMA_VERSION)
            .migration(MyMigration())
            .compactOnLaunch(object : CompactOnLaunchCallback {
                override fun shouldCompact(totalBytes: Long, usedBytes: Long): Boolean {
                    val thresholdSize = (50 * 1024 * 1024).toLong()
                    return totalBytes > thresholdSize && usedBytes.toDouble() / totalBytes.toDouble() < 0.5
                }
                override fun hashCode() = 9
                override fun equals(other: Any?): Boolean = true
            })
        }
        builder.build()
    }


SyncAdapter:

    override fun onPerformSync(account: Account, extras: Bundle, authority: String, provider: ContentProviderClient, syncResult: SyncResult) {
        // Initialization
        val realm = Realm.getInstance(realmConfig)
        // All the sync operations
        realm.close()
    }

What are the reasons you're talking about that can cause the crash? Maybe it can help us for the investigation.

**Comment by MariusBoepple** [comment]
It did occur again on a Pixel 2 Android 8.1:

```
04-26 16:21:29.604 W/SyncAdapterThre: type=1400 audit(0.0:7723): avc: denied { write } for name="perfd" dev="sda45" ino=1572867 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:shell_data_file:s0 tclass=dir permissive=0
    type=1400 audit(0.0:7724): avc: denied { write } for name="perfd" dev="sda45" ino=1572867 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:shell_data_file:s0 tclass=dir permissive=0
04-26 16:21:29.618 V/ExampleApp: Compressed network request from 2780 to 712, archiving a 0.256115 compression rate
04-26 16:21:29.623 D/ExampleApp: --> POST https://example.net/trips/7c0cc97f-1950-43f8-bc51-76ce226b1dbd/samples (712-byte body)
04-26 16:21:29.879 D/ExampleApp: <-- 204 No Content https://example.net/trips/7c0cc97f-1950-43f8-bc51-76ce226b1dbd/samples (254ms, 0-byte body)

04-26 16:21:29.898 E/REALM_JNI: jni: ThrowingException 5, flock() failed: Bad file descriptor in /Users/Nabil/Dev/realm/master/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_OsSharedRealm.cpp line 126, .
    Exception has been thrown: Unrecoverable error. flock() failed: Bad file descriptor in /Users/Nabil/Dev/realm/master/realm-java/realm/realm-library/src/main/cpp/io_realm_internal_OsSharedRealm.cpp line 126
04-26 16:21:29.900 E/REALM: /home/jenkins/workspace/lm_realm-core_release_5.4.2-7MRDQ2N52D4436PA46O7GFYLIKQGV37WCE5DFMFVZ5J6WJYDBC3A/src/realm/util/file.cpp:427: [realm-core-5.4.2] Assertion failed: r == 0
    !!! IMPORTANT: Please send this log and info about Realm SDK version and other relevant reproduction info to help@realm.io.

04-26 16:21:29.900 A/libc: Fatal signal 6 (SIGABRT), code -6 in tid 22577 (SyncAdapterThre), pid 18035 (example.app)
04-26 16:21:29.954 W/crash_dump64: type=1400 audit(0.0:7725): avc: denied { search } for name="com.example.app" dev="sda45" ino=2883844 scontext=u:r:crash_dump:s0:c512,c768 tcontext=u:object_r:app_data_file:s0:c512,c768 tclass=dir permissive=0
04-26 16:21:29.977 I/crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstone
04-26 16:21:29.983 I//system/bin/tombstoned: received crash request for pid 18035
04-26 16:21:29.987 I/crash_dump64: performing dump of process 18035 (target tid = 22577)
04-26 16:21:29.988 A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
    Build fingerprint: 'google/walleye/walleye:8.1.0/OPM2.171019.029/4657601:user/release-keys'
    Revision: 'MP1'
    ABI: 'arm64'
    pid: 18035, tid: 22577, name: SyncAdapterThre  >>> com.example.app <<<
    signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
        x0   0000000000000000  x1   0000000000005831  x2   0000000000000006  x3   0000000000000008
        x4   fefeff7742cf8487  x5   fefeff7742cf8487  x6   fefeff7742cf8487  x7   7f7f7f7f7fffffff
        x8   0000000000000083  x9   0000000010000000  x10  0000007843d05a20  x11  0000000000000001
        x12  0000007846d7db98  x13  ffffffffffffffff  x14  ffffffffff000000  x15  ffffffffffffffff
        x16  0000005b3aa09fa8  x17  00000078e4bc752c  x18  00000000ebad6082  x19  0000000000004673
        x20  0000000000005831  x21  0000000000000083  x22  000000784c23ac9b  x23  00000078620715b0
        x24  00000078620706e0  x25  0000007843d05d68  x26  0000000000000000  x27  0000000000000002
        x28  0000000000000001  x29  0000007843d05a60  x30  00000078e4b7c760
        sp   0000007843d05a20  pc   00000078e4b7c788  pstate 0000000060000000
04-26 16:21:29.992 A/DEBUG: backtrace:
        #00 pc 000000000001d788  /system/lib64/libc.so (abort+120)
        #01 pc 0000000000204b68  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #02 pc 0000000000204bac  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #03 pc 0000000000204c3c  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #04 pc 0000000000200b4c  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #05 pc 000000000019debc  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #06 pc 00000000000ea424  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #07 pc 000000000019fa78  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #08 pc 0000000000104424  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #09 pc 00000000000e9384  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so
        #10 pc 000000000006a36c  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so (Java_io_realm_internal_OsSharedRealm_nativeCloseSharedRealm+52)
        #11 pc 0000000000036a08  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/oat/arm64/base.odex (offset 0x2e000)
```
For the realm operations, I am using this extension:
```
internal fun <T> database(block: Realm.() -> T): T {
    Realm.getInstance(realmConfig).use {
        it.refresh()
        return block(it)
    }
}
```

The place where the crash probably occurred (based on the API calls from the log):

```
private fun uploadProtobufSamples() {
        database {
            // Get IDs of the sessions which have samples.
            val distinctSamples = where<Samples>()
                    .distinct(Samples.SESSION_UUID)
                    .findAll()

            // Loop over every session and upload the samples.
            for (distinctSample in distinctSamples) {
                val sessionUuid = distinctSample.sessionUuid

                // Get all samples which belong to that session.
                val sessionsSamples = where<Samples>()
                        .equalTo(Samples.SESSION_UUID, sessionUuid)
                        .findAll()

                // Set up protobuf builder.
                val processedSamples = ArrayList<Samples>()
                val uploadedSamples = ArrayList<Samples>()
                val builder = Builder()

                // Loop through all sample containers.
                for (samples in sessionsSamples) {

                    // Add samples container to the builder.
                    try {
                        builder.session_samples.addAll(com.example.app.buffers.Samples.ADAPTER.decode(samples.data).session_samples)

                    } catch (e: IOException) {
                        Log.e(e, "Dropping samples because decoding failed")
                        executeTransaction { samples.deleteFromRealm() }
                        continue
                    }

                    // Mark container as processed.
                    processedSamples.add(samples)

                    // If there are more than x samples in the builder, upload the batch.
                    if (builder.trip_samples.size > PROTO_BATCH_SIZE_LIMIT) {
                        // HERE THE LOGS ARE PRINTED:
                        val successful = uploadSamplesBatch(sessionUuid, builder)

                        if (successful) {
                            uploadedSamples.addAll(processedSamples)
                        }

                        // Clear builder and list to make a new batch.
                        builder.trip_samples.clear()
                        processedSamples.clear()
                    }
                }

                // Upload the remaining samples...

                // Delete all the uploaded samples.
                executeTransaction {
                    for (samples in uploadedSamples) {
                        samples.deleteFromRealm()
                    }
                }
            }
        }
    }
```
**TL;DR**
At this place, some objects from a `RealmResults` are added to `ArrayLists` and then deleted.

**Comment by Zhuinden** [comment]
I really think the Realm is deleted while the operations occur.

**Comment by cmelchior** [comment]
Unstripped dump:
```
********** Crash dump: **********
Build fingerprint: 'google/walleye/walleye:8.1.0/OPM2.171019.029/4657601:user/release-keys'
pid: 18035, tid: 22577, name: SyncAdapterThre  >>> com.example.app <<<
signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
Stack frame #00 pc 000000000001d788  /system/lib64/libc.so (abort+120)
Stack frame #01 pc 0000000000204b68  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine please_report_this_error_to_help_at_realm_dot_io at :?
Stack frame #02 pc 0000000000204bac  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine realm::util::terminate_internal(std::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&) at :?
Stack frame #03 pc 0000000000204c3c  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine realm::util::terminate(char const*, char const*, long, std::initializer_list<realm::util::Printable>&&) at :?
Stack frame #04 pc 0000000000200b4c  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine realm::util::File::close() at :?
Stack frame #05 pc 000000000019debc  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine std::_Sp_counted_ptr_inplace<realm::util::InterprocessMutex::LockInfo, std::allocator<realm::util::InterprocessMutex::LockInfo>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() at :?
Stack frame #06 pc 00000000000ea424  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine realm::util::InterprocessMutex::free_lock_info() at :?
Stack frame #07 pc 000000000019fa78  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine realm::SharedGroup::~SharedGroup() at :?
Stack frame #08 pc 0000000000104424  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine realm::_impl::RealmCoordinator::~RealmCoordinator() at :?
Stack frame #09 pc 00000000000e9384  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so: Routine realm::Realm::close() at :?
Stack frame #10 pc 000000000006a36c  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/lib/arm64/librealm-jni.so (Java_io_realm_internal_OsSharedRealm_nativeCloseSharedRealm+52): Routine Java_io_realm_internal_OsSharedRealm_nativeCloseSharedRealm at :?
Stack frame #11 pc 0000000000036a08  /data/app/com.example.app-C37EAX8h9yrQZolgTquoAw==/oat/arm64/base.odex (offset 0x2e000)
```

**Comment by cmelchior** [comment]
Hi @MariusBoepple Thank you very much for this. I'll run this past our Core people in order to determine if this can happen in other scenarios than accidentially deleting the Realm file
